orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/CFRollingRedBlackStrategy.java

```diff
@@ -0,0 +1,290 @@
+    try {
+      source = getSource(targetServerGroupResolver, stageData, baseContext);
+    } catch (Exception e) {
+      source = null;
+    }
+
+    if (source == null) {
+      log.warn("no source server group -- will perform RRB to exact fallback capacity " + savedCapacity + " with no disableCluster or scaleDownCluster stages");
+    }
+
+    ResizeStrategy.Capacity sourceCapacity = source == null ?
+      savedCapacity :
+      resizeStrategySupport.getCapacity(source.getCredentials(), source.getServerGroupName(), source.getCloudProvider(), source.getLocation());
+
+    for (Integer percentage : targetPercentages) {
+      Map<String, Object> scaleUpContext = getScalingContext(stage, cleanupConfig, baseContext, savedCapacity, percentage, null);
+
+      log.info("Adding `Grow target to " + (percentage) + "% of desired size` stage with context " + scaleUpContext + " [executionId=" + executionId + "]");
+
+      Stage resizeStage = new Stage(execution, ResizeServerGroupStage.TYPE, "Grow target to " + percentage + "% of desired size", scaleUpContext);
+      resizeStage.setParentStageId(stage.getId());
+      resizeStage.setSyntheticStageOwner(SyntheticStageOwner.STAGE_AFTER);
+      stages.add(resizeStage);
+
+      // only generate the "disable p% of traffic" stages if we have something to disable
+      if (source != null) {
+        stages.addAll(getBeforeCleanupStages(stage, stageData));
+        Map<String, Object> scaleDownContext =
+          getScalingContext(stage, cleanupConfig, baseContext, sourceCapacity, 100 - percentage, source.getServerGroupName());
+        scaleDownContext.put("scaleStoppedServerGroup", true);
+
+        log.info("Adding `Shrink source to " + (100 - percentage) + "% of initial size` stage with context " + scaleDownContext + " [executionId=" + executionId + "]");
+
+        Stage scaleDownStage = new Stage(execution, ResizeServerGroupStage.TYPE, "Shrink source to " + (100 - percentage) + "% of initial size", scaleDownContext);
+        scaleDownStage.setParentStageId(stage.getId());
+        scaleDownStage.setSyntheticStageOwner(SyntheticStageOwner.STAGE_AFTER);
+        stages.add(scaleDownStage);
+      }
+    }
+
+    if (source != null) {
+      // shrink cluster to size
+      Map<String, Object> shrinkClusterContext = new HashMap<>(baseContext);
+      shrinkClusterContext.put("allowDeleteActive", false);
+      shrinkClusterContext.put("shrinkToSize", stage.getContext().get("maxRemainingAsgs"));
+      shrinkClusterContext.put("retainLargerOverNewer", false);
+      Stage shrinkClusterStage = new Stage(execution, ShrinkClusterStage.STAGE_TYPE, "shrinkCluster", shrinkClusterContext);
+      shrinkClusterStage.setParentStageId(stage.getId());
+      shrinkClusterStage.setSyntheticStageOwner(SyntheticStageOwner.STAGE_AFTER);
+      stages.add(shrinkClusterStage);
+
+      // disable old
+      log.info("Adding `Disable cluster` stage with context " + baseContext + " [executionId=" + executionId + "]");
```

Comment:
For slf4j, please use `{}` + varargs rather than string concatenation.

Reply:
Done!